src/plugins/dashboard/public/application/embeddable/viewport/dashboard_viewport.tsx

```diff
@@ -104,8 +98,38 @@ export class DashboardViewport extends React.Component<DashboardViewportProps, S
     });
   };
 
+  private shouldRenderEmpty() {
+    const container = this.props.container;
+    const getShouldShowEditHelp = () =>
+      container.getPanelCount() === 0 &&
+      container.getInput().viewMode !== ViewMode.VIEW &&
+      !container.getInput().dashboardCapabilities?.hideWriteControls;
+
+    const getShouldShowViewHelp = () =>
+      container.getPanelCount() === 0 &&
+      container.getInput().viewMode === ViewMode.VIEW &&
+      !container.getInput().dashboardCapabilities?.hideWriteControls;
+
+    const shouldShowUnauthorizedEmptyState = () => {
+      const readonlyMode =
+        container.getPanelCount() === 0 &&
+        !getShouldShowEditHelp() &&
+        !getShouldShowViewHelp() &&
+        container.getInput().dashboardCapabilities?.hideWriteControls;
+      const userHasNoPermissions =
+        container.getPanelCount() === 0 &&
+        !container.getInput().dashboardCapabilities?.visualizeCapabilities.save &&
+        !container.getInput().dashboardCapabilities?.mapsCapabilities.save;
+      return readonlyMode || userHasNoPermissions;
+    };
+
+    const shouldShowEditHelp = getShouldShowEditHelp();
+    const shouldShowViewHelp = getShouldShowViewHelp();
+    const isEmptyInReadOnlyMode = shouldShowUnauthorizedEmptyState();
```

Comment:
You're repeating calls to a lot of methods, and destructuring would save you a lot of repetition, as well.

Case in point: I just started to destructure this, and noticed--

```ts
const getShouldShowEditHelp = () =>
  container.getPanelCount() === 0 &&
  container.getInput().viewMode !== ViewMode.VIEW &&
  !container.getInput().dashboardCapabilities?.hideWriteControls;

const getShouldShowViewHelp = () =>
  container.getPanelCount() === 0 &&
  container.getInput().viewMode === ViewMode.VIEW &&
  !container.getInput().dashboardCapabilities?.hideWriteControls;
```

-- are **identical**.  The fact they're long makes it too easy to skip past them.

Here is the result of destructuring/deduplicating:

```ts
  const shouldRenderEmpty = () => {
    const container = this.props.container;

    const panelCount = container.getPanelCount();
    const input = container.getInput();

    const { viewMode, dashboardCapabilities } = input;
    const { hideWriteControls, visualizeCapabilities, mapsCapabilities } = dashboardCapabilities || {};

    const isReadyOnly = 
      panelCount === 0 &&
      viewMode !== ViewMode.VIEW &&
      !hideWriteControls;
    
    const hasNoPermission = 
      panelCount === 0 &&
      !visualizeCapabilities?.save &&
      !mapsCapabilities?.save;
    
    return isReadyOnly || hasNoPermission;
  };
```

Reply:
I hadn't actually looked into this code - just copied it from another place. But on closer inspection, it appears like that entire method can actually be replaced with `this.props.container.getPanelCount() === 0`